crates/ide/src/completion/complete_mod.rs

```diff
@@ -0,0 +1,317 @@
+            fn foo() {}
+            //- /foo/ignored_foo.rs
+            fn ignored_foo() {}
+            //- /bar/mod.rs
+            fn bar() {}
+            //- /bar/ignored_bar.rs
+            fn ignored_bar() {}
+        "#,
+            expect![[r#"
+                md bar;
+                md foo;
+            "#]],
+        );
+    }
+
+    #[test]
+    fn no_module_completion_with_module_body() {
+        check(
+            r#"
+            //- /lib.rs
+            mod <|> {
+
+            }
+            //- /foo.rs
+            fn foo() {}
+        "#,
+            expect![[r#""#]],
+        );
+    }
+
+    #[test]
+    fn main_module_completion() {
+        check(
+            r#"
+            //- /main.rs
+            mod <|>
+            //- /foo.rs
+            fn foo() {}
+            //- /foo/ignored_foo.rs
+            fn ignored_foo() {}
+            //- /bar/mod.rs
+            fn bar() {}
+            //- /bar/ignored_bar.rs
+            fn ignored_bar() {}
+        "#,
+            expect![[r#"
+                md bar;
+                md foo;
+            "#]],
+        );
+    }
+
+    #[test]
+    fn main_test_module_completion() {
+        check(
+            r#"
+            //- /main.rs
+            mod tests {
+                mod <|>;
+            }
+            //- /tests/foo.rs
+            fn foo() {}
+        "#,
+            expect![[r#"
+                md foo
+            "#]],
+        );
+    }
+
+    #[test]
+    fn directly_nested_module_completion() {
+        check(
+            r#"
+            //- /lib.rs
+            mod foo;
+            //- /foo.rs
+            mod <|>;
+            //- /foo/bar.rs
+            fn bar() {}
+            //- /foo/bar/ignored_bar.rs
+            fn ignored_bar() {}
+            //- /foo/baz/mod.rs
+            fn baz() {}
+            //- /foo/moar/ignored_moar.rs
+            fn ignored_moar() {}
+        "#,
+            expect![[r#"
+                md bar
+                md baz
+            "#]],
+        );
+    }
+
+    #[test]
+    fn nested_in_source_module_completion() {
+        check(
+            r#"
+            //- /lib.rs
+            mod foo;
+            //- /foo.rs
+            mod bar {
+                mod <|>
+            }
+            //- /foo/bar/baz.rs
+            fn baz() {}
+        "#,
+            expect![[r#"
+                md baz;
+            "#]],
+        );
+    }
+
+    // FIXME binart modules are not picked up in tests
+    // #[test]
+    // fn regular_bin_module_completion() {
+    //     check(
+    //         r#"
+    //         //- /src/main.rs
+    //         fn main() {}
+    //         //- /src/main/foo.rs
+    //         mod <|>
+    //         //- /src/main/bar.rs
+    //         fn bar() {}
+    //         //- /src/main/bar/bar_ignored.rs
+    //         fn bar_ignored() {}
+    //     "#,
+    //         expect![[r#"
+    //             md bar;
+    //         "#]],
+    //     );
+    // }
```

Comment:
What is this test trying to test? By "binary module" do you mean Cargo's automatic binary crates in `src/bin`?

Reply:
Yes, they are able to import modules from `src/bin`, so are a bit exceptional in the way we should look for importable modules.

Reply:
I think they're just normal crates, no? If the crate's root file is at `src/bin/x.rs`, then it should be able to import other files in `src/lib` via `mod file;`, just like other root files like `src/lib.rs` can inside `lib`.

Reply:
I don't think so, or I'm doing something wrong.
Judging from the compiler errors, modules in `src/bin` are able to import any other non-imported module from `src/bin` and nothing else:


```
someonetoignore at alaptop in ~/Downloads/tmp_dir/zzzz
$ tree .
.
├── Cargo.lock
├── Cargo.toml
└── src
    ├── bin
    │   ├── bin_1.rs
    │   └── bin_module.rs
    ├── lib.rs
    └── test_lib_module.rs
```

`lib.rs` can contain `mod test_lib_module;` only
`bin.rs` can contain `mod bin_module; `only:

<img width="1438" alt="image" src="https://user-images.githubusercontent.com/2690773/92868066-ffb29c80-f409-11ea-98a4-4bc12d3dc943.png">

When I add `mod test_lib_module;` into `bin_1.rs`, I get a compiler error:
```
error[E0583]: file not found for module `test_lib_module`
 --> src/bin/bin_1.rs:2:1
  |
2 | mod test_lib_module;
  | ^^^^^^^^^^^^^^^^^^^^
  |
  = help: to create the module `test_lib_module`, create file "src/bin/test_lib_module.rs"

error: aborting due to previous error
```

Reply:
Yeah, that's what I would expect. If you invoke `rustc path/to/file.rs`, then `file.rs` can only declare modules that are inside `path/to/`, not `path/` itself. This is the same for library and binary crates.

That's why I'm confused as to why this would require extra tests, since it seems consistent to me. Or is the only thing that needs testing here that a main file in a folder won't include modules in the parent directory?

Reply:
That plus checking that we're not allowed to create cycles with mod declarations.

But if this test confuses you so much, I can delete it.

Reply:
No, I was just trying to understand what it is meant to do, so maybe clarify the comment a bit? Then it should be fine to land.

Reply:
Done.